File: mac/VoiceTests/HotkeyInterpreterTests.swift
import XCTest
@testable import Voice

final class HotkeyInterpreterTests: XCTestCase {
    func testFnHoldAndRelease() {
        var i = HotkeyInterpreter(choice: .fn)
        XCTAssertEqual(i.handle(.flags(fn: true, rightOption: false, rightCommand: false)), .press)
        XCTAssertEqual(i.handle(.flags(fn: false, rightOption: false, rightCommand: false)), .release)
    }

    func testOtherKeyWhileHeldCancelsAndSwallowsTheRelease() {
        var i = HotkeyInterpreter(choice: .fn)
        _ = i.handle(.flags(fn: true, rightOption: false, rightCommand: false))
        XCTAssertEqual(i.handle(.keyDown(96)), .cancel)           // F5
        XCTAssertNil(i.handle(.flags(fn: false, rightOption: false, rightCommand: false)))
        XCTAssertEqual(i.handle(.flags(fn: true, rightOption: false, rightCommand: false)), .press)
    }

    func testEscapeCancels() {
        var i = HotkeyInterpreter(choice: .fn)
        _ = i.handle(.flags(fn: true, rightOption: false, rightCommand: false))
        XCTAssertEqual(i.handle(.keyDown(53)), .cancel)
    }

    func testKeysWhileIdleAreIgnored() {
        var i = HotkeyInterpreter(choice: .fn)
        XCTAssertNil(i.handle(.keyDown(0)))
        XCTAssertNil(i.handle(.flags(fn: false, rightOption: true, rightCommand: false)))
    }

    func testRightOptionChoiceIgnoresFn() {
        var i = HotkeyInterpreter(choice: .rightOption)
        XCTAssertNil(i.handle(.flags(fn: true, rightOption: false, rightCommand: false)))
        XCTAssertEqual(i.handle(.flags(fn: false, rightOption: true, rightCommand: false)), .press)
        XCTAssertEqual(i.handle(.flags(fn: false, rightOption: false, rightCommand: false)), .release)
    }

    // D2: isHeld is true after press, false after release — HotkeyMonitor.setChoice relies on this
    // to decide whether a Settings hotkey change must be deferred to the next release/cancel.
    func testIsHeldReflectsPressAndRelease() {
        var i = HotkeyInterpreter(choice: .fn)
        XCTAssertFalse(i.isHeld)
        _ = i.handle(.flags(fn: true, rightOption: false, rightCommand: false))
        XCTAssertTrue(i.isHeld)
        _ = i.handle(.flags(fn: false, rightOption: false, rightCommand: false))
        XCTAssertFalse(i.isHeld)
    }

    // H1: press → Esc (cancel) resets isHeld immediately — it must not wait for the physical
    // flags-release that follows, which is swallowed (returns nil) and produces no further action.
    func testIsHeldResetsOnCancelBeforeThePhysicalRelease() {
        var i = HotkeyInterpreter(choice: .fn)
        _ = i.handle(.flags(fn: true, rightOption: false, rightCommand: false))
        XCTAssertTrue(i.isHeld)
        XCTAssertEqual(i.handle(.keyDown(HotkeyInterpreter.escapeKeyCode)), .cancel)
        XCTAssertFalse(i.isHeld)
        XCTAssertNil(i.handle(.flags(fn: false, rightOption: false, rightCommand: false)))
        XCTAssertFalse(i.isHeld)
    }

    // MARK: - Latched mode (prd-hands-free-dictation.md rule 6)

    func testALatchedSessionIgnoresOrdinaryKeystrokes() {
        // The entire point of hands-free is that the user keeps typing. Before this, any keyDown
        // during a dictation cancelled it, so a latched session would have died on the first
        // keystroke — the one thing latched mode exists to allow.
        var i = HotkeyInterpreter(choice: .fn)
        i.latched = true
        for code: UInt16 in [0, 1, 36, 49, 123] {   // a, s, return, space, left-arrow
            XCTAssertNil(i.handle(.keyDown(code)), "keyCode \(code) must not cancel a latched session")
        }
    }

    func testEscStillCancelsALatchedSession() {
        // The single exception. Once the key is no longer held there is no other way to say
        // "throw this away" — and a 30-second hands-free session is exactly when you want one.
        var i = HotkeyInterpreter(choice: .fn)
        i.latched = true
        XCTAssertEqual(i.handle(.keyDown(HotkeyInterpreter.escapeKeyCode)), .cancel)
    }

    func testUnlatchedBehaviourIsUnchanged() {
        // The hold path must be untouched: during a hold, any key is a shortcut and cancels.
        var i = HotkeyInterpreter(choice: .fn)
        XCTAssertEqual(i.handle(.flags(fn: true, rightOption: false, rightCommand: false)), .press)
        XCTAssertEqual(i.handle(.keyDown(0)), .cancel)
    }
}

File: windows/Spit.Core.Tests/HotkeyInterpreterTests.cs
namespace Spit.Core.Tests;

// Port of mac/VoiceTests/HotkeyInterpreterTests.swift with Windows keys (rule 27). Names are the
// Mac's: Fn (the Mac default) maps to Right Ctrl (the Windows default), Right Option to Right Alt, and
// macOS key codes to virtual-key codes (F5 96 → 0x74, Esc 53 → 0x1B, A 0 → 0x41, S 1 → 0x53,
// Return 36 → 0x0D, Space 49 → 0x20, Left 123 → 0x25).
public sealed class HotkeyInterpreterTests
{
    private static KeyEvent Flags(bool rightCtrl = false, bool rightAlt = false) => new KeyEvent.Flags(rightCtrl, rightAlt);
    private static KeyEvent KeyDown(int vk) => new KeyEvent.KeyDown(vk);

    [Fact]
    public void testFnHoldAndRelease()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        Assert.Equal(HotkeyAction.Press, i.Handle(Flags(rightCtrl: true)));
        Assert.Equal(HotkeyAction.Release, i.Handle(Flags()));
    }

    [Fact]
    public void testOtherKeyWhileHeldCancelsAndSwallowsTheRelease()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        _ = i.Handle(Flags(rightCtrl: true));
        Assert.Equal(HotkeyAction.Cancel, i.Handle(KeyDown(0x74)));   // F5
        Assert.Null(i.Handle(Flags()));
        Assert.Equal(HotkeyAction.Press, i.Handle(Flags(rightCtrl: true)));
    }

    [Fact]
    public void testEscapeCancels()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        _ = i.Handle(Flags(rightCtrl: true));
        Assert.Equal(HotkeyAction.Cancel, i.Handle(KeyDown(0x1B)));
    }

    [Fact]
    public void testKeysWhileIdleAreIgnored()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        Assert.Null(i.Handle(KeyDown(0x41)));
        Assert.Null(i.Handle(Flags(rightAlt: true)));
    }

    [Fact]
    public void testRightOptionChoiceIgnoresFn()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightAlt);
        Assert.Null(i.Handle(Flags(rightCtrl: true)));
        Assert.Equal(HotkeyAction.Press, i.Handle(Flags(rightAlt: true)));
        Assert.Equal(HotkeyAction.Release, i.Handle(Flags()));
    }

    // D2: IsHeld is true after press, false after release — the monitor relies on this to decide
    // whether a Settings hotkey change must be deferred to the next release/cancel.
    [Fact]
    public void testIsHeldReflectsPressAndRelease()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        Assert.False(i.IsHeld);
        _ = i.Handle(Flags(rightCtrl: true));
        Assert.True(i.IsHeld);
        _ = i.Handle(Flags());
        Assert.False(i.IsHeld);
    }

    // H1: press → Esc (cancel) resets IsHeld immediately — it must not wait for the physical release
    // that follows, which is swallowed and produces no further action.
    [Fact]
    public void testIsHeldResetsOnCancelBeforeThePhysicalRelease()
    {
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        _ = i.Handle(Flags(rightCtrl: true));
        Assert.True(i.IsHeld);
        Assert.Equal(HotkeyAction.Cancel, i.Handle(KeyDown(HotkeyInterpreter.EscapeKeyCode)));
        Assert.False(i.IsHeld);
        Assert.Null(i.Handle(Flags()));
        Assert.False(i.IsHeld);
    }

    // MARK: - Latched mode (prd-hands-free-dictation.md rule 6)

    [Fact]
    public void testALatchedSessionIgnoresOrdinaryKeystrokes()
    {
        // The entire point of hands-free is that the user keeps typing.
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl) { Latched = true };
        foreach (var vk in new[] { 0x41, 0x53, 0x0D, 0x20, 0x25 })   // a, s, return, space, left-arrow
        {
            Assert.Null(i.Handle(KeyDown(vk)));   // must not cancel a latched session
        }
    }

    [Fact]
    public void testEscStillCancelsALatchedSession()
    {
        // The single exception: once the key is no longer held there is no other way to say "throw this away".
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl) { Latched = true };
        Assert.Equal(HotkeyAction.Cancel, i.Handle(KeyDown(HotkeyInterpreter.EscapeKeyCode)));
    }

    [Fact]
    public void testUnlatchedBehaviourIsUnchanged()
    {
        // The hold path must be untouched: during a hold, any key is a shortcut and cancels.
        var i = new HotkeyInterpreter(HotkeyChoice.RightCtrl);
        Assert.Equal(HotkeyAction.Press, i.Handle(Flags(rightCtrl: true)));
        Assert.Equal(HotkeyAction.Cancel, i.Handle(KeyDown(0x41)));
    }
}
